### PR TITLE
Make array types implement ArrayAccess and Iterator

### LINKED ISSUE #168: Make array types implement ArrayAccess, Iterator
After reintroducing Array classes (#160) we could improve developer experience by making it easier to work with these types. This could be achieved my making them extend the [`ArrayAccess`](http://php.net/manual/en/class.arrayaccess.php) and/or [`Iterator`](http://php.net/manual/en/class.iterator.php) interfaces.

One suggested approach (#110) for identifying array types is:

> types that have a sequence with exacly one element and cardinality for that element is either:
> - n...unbounded (where n is >= 0)
> - or - n...m (where n >=0 and m > 1)

### PR DESCRIPTION
**Pull request by red-led**
Closes #168

### CODE REVIEW

**Comment by kasperg** [comment]
Cheers! Happy to see work being done here.

**Comment by red-led** [comment]
Up! Fixed error...

**Comment by kasperg** [comment]
Sorry for the late response - just got back from summer vacation. I am going to take a look at this the next couple of days.

File: src/Xml/TypeNode.php [hunk 1]
```diff
@@ -230,6 +220,28 @@ public function isComplex()
...
+            || count($parts) !== 1 || (
+                substr($this->name, -2, 2) !== '[]' &&
+                substr($this->name, 0, 7) !== 'ArrayOf')
+        ) {
+            return false;
+        }
+
+        $varType = reset($parts);
+
+        return substr($varType, -2, 2) == '[]';
```
Comment (kasperg) [inline src/Xml/TypeNode.php]: This change increases the complexity of the `isArray()` method quite a bit. I would appreciate an inline comment explaining why such a check is necessary.

File: src/ArrayType.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+use Wsdl2PhpGenerator\PhpSource\PhpDocComment;
+use Wsdl2PhpGenerator\PhpSource\PhpDocElementFactory;
+use Wsdl2PhpGenerator\PhpSource\PhpFunction;
+
+/**
+ * ArrayType
+ *
+ * @package Wsdl2PhpGenerator
+ * @author Fredrik Wallgren <fredrik.wallgren@gmail.com>
+ * @license http://www.opensource.org/licenses/mit-license.php MIT License
```
Comment (kasperg) [inline src/ArrayType.php]: Please remove the `@author` and `@license` annotations.

This has a been community project for a while even though there a similar traces of this throughout existing classes. The license information is included in the `LICENSE` file.

File: src/ArrayType.php [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+ */
+class ArrayType extends ComplexType
+{
+    /** @var Variable */
+    protected $field;
+
+    /** @var string */
```
Comment (kasperg) [inline src/ArrayType.php]: Please add a line or two describing what these properties represent.

**Comment by kasperg** [comment]
Thanks for the suggestion @red-led. Again: Apologies for the late review.

The implementation looks solid and even though I normally discourage changes to `PhpSource` classes I am quite confident that moving to a separate library would support implementation of interfaces just as well.

I would appreciate a unit test for this. Would you look into that?

Also the pull request no longer applies cleanly. Please rebase.

**Comment by red-led** [comment]
Hi @kasperg! Now it's my time for summer vacation, so it will take some time to make changes.

**Comment by kasperg** [comment]
No problem :).

**Comment by red-led** [comment]
Three months... and I'v just reminded about this :persevere:
Anyway, done.

**Comment by red-led** [comment]
Thank's @nkm for finding some bugs and unit tests implementation!

**Comment by nkm** [comment]
Thank you guys for your work :)

File: src/ComplexType.php [hunk 4]
```diff
@@ -65,17 +65,9 @@ protected function generateClass()
...
-        // It is actually possible to have different classes with the same name as PHP SoapClient has a poor
-        // understanding of namespaces. Two types with the same name but in different namespaces will have the same
-        // identifier.
-        if ($this->baseType !== null && $this->baseType !== $this) {
-            $classBaseType = $this->baseType->getPhpIdentifier();
-        }
+        $classBaseType = $this->getBaseTypeClass();
 
-        $class = new PhpClass(
+        $this->class = new PhpClass(
```
Comment (kasperg) [inline src/ComplexType.php]: `$class` is now a class property. Please declare it.

File: src/ComplexType.php [hunk 5]
```diff
@@ -194,13 +184,28 @@ protected function generateClass()
...
 
         foreach ($accessors as $accessor) {
-            $class->addFunction($accessor);
+            $this->class->addFunction($accessor);
+        }
+    }
+
+    /**
+     * Determine parent class
+     * @return string
```
Comment (kasperg) [inline src/ComplexType.php]: This returns a string containing the PHP identifier for the parent class or `null` if there is no applicable parent class.

File: src/Xml/TypeNode.php [hunk 6]
```diff
@@ -230,6 +220,29 @@ public function isComplex()
...
+            || count($parts) !== 1
+            || (substr($this->name, 0, 7) !== 'ArrayOf')
+        ) {
+            return false;
+        }
+
+        $varType = reset($parts);
+
+        // Check if only field is array
+        return substr($varType, -2, 2) == '[]';
```
Comment (kasperg) [inline src/Xml/TypeNode.php]: Could we simplify this function as so:

```php
$parts = $this->getParts();
// Array types are complex types with one element, 
//the array name begins with 'ArrayOf' and the element is an array.
return ($this->isComplex()
	&& count($parts) == 1
 	&& (substr($this->name, 0, 7) !== 'ArrayOf')
 	&& substr(reset($parts), -2, 2) == '[]';
```

**Comment by kasperg** [comment]
@red-led: I have re-reviewed this. This still looks very good :+1:

I have a couple of minor commets I would appreciate if you could look into.

Once addressed please consider if this could be squashed into fewer commits even if this is the work of multiple developers. I will make sure @nkm is mentioned in the list of contributors.

File: src/ComplexType.php (hunk 4 above)
Comment (red-led) [inline src/ComplexType.php]: It is declared in `\Wsdl2PhpGenerator\Type`:
```
    /**
     * @var PhpClass The class used to create the type. This is not used by patterns
     */
    protected $class;
```

Should I re-declare it?

Comment (kasperg) [inline src/ComplexType.php]: No - my mistake not spotting it up the inheritance chain.

**Comment by red-led** [comment]
@kasperg: done. I have also saved @nkm's authorship on testcase implementation : )

**Comment by kasperg** [comment]
Great. Merged. Thanks a lot for both of your work.

**Pull request merged**